raiden/network/proxies/service_registry.py

```diff
@@ -0,0 +1,85 @@
+import structlog
+import web3
+from eth_utils import is_binary_address, to_normalized_address
+
+from raiden.exceptions import InvalidAddress
+from raiden.network.proxies.utils import compare_contract_versions
+from raiden.network.rpc.client import JSONRPCClient, check_address_has_code
+from raiden.network.rpc.transactions import check_transaction_threw
+from raiden.utils.typing import Address, AddressHex, BlockSpecification, Optional
+from raiden_contracts.constants import CONTRACT_SERVICE_REGISTRY
+from raiden_contracts.contract_manager import ContractManager
+
+log = structlog.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+class ServiceRegistry:
+    def __init__(
+            self,
+            jsonrpc_client: JSONRPCClient,
+            service_registry_address: Address,
+            contract_manager: ContractManager,
+    ):
+        if not is_binary_address(service_registry_address):
+            raise InvalidAddress('Expected binary address for service registry')
+
+        self.contract_manager = contract_manager
+        check_address_has_code(jsonrpc_client, service_registry_address, CONTRACT_SERVICE_REGISTRY)
+
+        proxy = jsonrpc_client.new_contract_proxy(
+            self.contract_manager.get_contract_abi(CONTRACT_SERVICE_REGISTRY),
+            to_normalized_address(service_registry_address),
+        )
+
+        compare_contract_versions(
+            proxy=proxy,
+            expected_version=contract_manager.contracts_version,
+            contract_name=CONTRACT_SERVICE_REGISTRY,
+            address=service_registry_address,
+        )
+
+        self.address = service_registry_address
+        self.proxy = proxy
+        self.client = jsonrpc_client
+        self.node_address = self.client.address
+
+    def service_count(self, block_identifier: BlockSpecification) -> int:
+        """Get the number of registered services"""
+        result = self.proxy.contract.functions.serviceCount().call(
+            block_identifier=block_identifier,
+        )
+        return result
+
+    def get_service_address(
+            self,
+            block_identifier: BlockSpecification,
+            index: int,
+    ) -> Optional[AddressHex]:
+        """Gets the address of a service by index. If index is out of range return None"""
+        try:
+            result = self.proxy.contract.functions.service_addresses(index).call(
+                block_identifier=block_identifier,
+            )
+        except web3.exceptions.BadFunctionCallOutput:
```

Comment:
Why is this exception being handled here but no in the `get_service_url` call?

Does the `web3` library behave differently depending on the type of the attribute?

```solidity
    mapping(address => string) public urls;  // URLs of services for HTTP access
    address[] public service_addresses; // list of available services (ethereum addresses)
```

Reply:
Yes. unknown key mapping to string will always return `''`.
But index out of range apparently throws BadFunctionCallOutput. There is a also a test for this.